keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -151,24 +153,14 @@ API server change:
     with the resources it can serve and receiving a request for a resource
     that is not yet available on that apiserver).
 
-* Discovery merging:
+* Discovery handling:
 
-  - During upgrade or downgrade, it may be the case that no apiserver has a
-    complete list of available resources. To fix the problems mentioned, it's
-    necessary that discovery exactly matches the capability of the system. So,
-    we will use the storage version objects to reconstruct a merged discovery
-    document and serve that in all apiservers.
+  - We will proxy the discovery request to the newest available apiserver in the cluster. In doing so, we will ensure that all discovery requests are handled keeping the future state of the cluster in mind. 
```

Comment:
I can't tell if this is proposing *always* redirecting discovery to the newest server, or only when servers disagree about the contents of discovery.

Reply:
I was thinking the former, i.e. always redirecting discovery to the latest apiserver. 
Thanks for clarifying my understanding of the GC and NLC behaviors. Yeah need to think more on how to handle the scenario where an API is not served by the latest apiserver.

Reply:
So, thinking more on how/whether to handle discovery requests as a part of UVIP I am seeing the following options:
a). implement a merged discovery doc as was originally proposed 
b). route discovery to newest apiserver
c). do nothing (drop discovery handling from UVIP)

Going by https://github.com/kubernetes/enhancements/pull/4441#discussion_r1467028748, the compatibility version feature  will make it possible to implement merging far more simply. With compatibility version we would be able to track historical API information, and if we use that to serve discovery, we also avoid the problem of GC and NLC becoming unaware of the previously-present-now-absent APIs (which happens if we route discovery to newest apiserver as a part of UVIP).

Maybe we just let discovery be as it is now and not address that for UVIP? Which means we avoid introducing significant changes that may not all make sense considering there will be future changes that compatibility versions will make to discovery. That would also avoid code complexity and developer time required to implement merging if we can get it mostly for free once compatibility version finishes.

TLDR: how about going with c). ?

Reply:
> Note also that the compatibility version KEP requires introducing the ability to produce a discovery document for a version different than the binary version

My initial assumption is this wouldn't require a code change. If compatibility version KEP is changing the storage version & enabling/disabling certain versions before API registration, then I'd expect that the discovery document should only reflect the installed/enabled APIs. 

I've seen controllers use the discovery information as the source of truth as to what GVs are available. IMO it would be surprising to be able to send a request to an apiserver and have it proxied/served, but that resource not be listed in discovery when I query the same service.

Reply:
So, once we make UVIP compatibility version aware, we will route client requests to only those apiservers that are compatible with the client. That is also one of the beta criteria added [here](https://github.com/kubernetes/enhancements/pull/4441/files#diff-5f6582f50a19d8d39a3a9045f6f9b6e404edd88ef829c187709437484bf308f8R212-R265). And the compatibility version KEP will provide discovery for a compatibility range. Agree that there will be some gap till when compatibility version aware discovery happens (UVIP should remain beta till then), but eventually we should be in a state where UVIP proxying and discovery are aligned.